lib/opam_solve.ml

```diff
@@ -817,11 +836,34 @@ module Mock_context :
                    (v, Error (UserConstraint (name, Some test)))
                | _ -> (v, Ok pkg))
 
-  type input = opam_env * OpamFile.OPAM.t list
+  type input = opam_env * OpamFile.OPAM.t list * OpamPackage.t list
+
+  let is_pinned t name = OpamPackage.has_name t.pins name
 
-  let create ?(test = OpamPackage.Name.Set.empty) ~constraints (env, pkgs) =
+  let create ?(test = OpamPackage.Name.Set.empty) ~constraints (env, pkgs, pins)
+      =
     let env varname = String.Map.find_opt varname env in
-    { pkgs; constraints; test; env }
+    let pins = OpamPackage.Set.of_list pins in
+    (* remove pinned packages from the universe -- as that's what's the
+       opam solver is doing. *)
+    let pp_pkg ppf pkg =
+      Fmt.pf ppf "%s.%s"
+        (OpamPackage.Name.to_string (OpamFile.OPAM.name pkg))
+        (OpamPackage.Version.to_string (OpamFile.OPAM.version pkg))
```

Comment:
I think `let pp_pkg = Fmt.of_to_string OpamPackage.to_string` can take care of this and will format it exactly in the `<name>.<version>` format.